### ISSUE TITLE
Separate server/user job-execution folders

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#473]
(based on several ongoing discussions, and edited accordingly)

---

# Files

Two kinds of files should be "exchanged" between server and cluster user

1. Files written by fractal (AKA the user running the server), from python:
    * JSON file with task input &#x1F534;
    * Sbatch scripts
    * Pickle file with cfut job input
    * Log files

2. Files written by test01
    * Log files for the tasks (written by test01, from python)
    * JSON file with task output (written by test01, from python)
    * SLURM stdout/stderr (written by SLURM)
    * Pickle file with cfut job output (written by test01, from python)

# sudo requirements:
* fractal can run `sudo -u test01 sbatch something.sh`
* fractal can run `sudo -u test01 cat something.sh`
* fractal can run `sudo -u test01 ls something.sh`

# Folder structure

### Current version
JOBDIR and USERJOBDIR are the same folder, owned by fractal and with 777 permissions. All the rest is the same as in future version.

### Future version:

Folders owned by fractal:
* ARTIFACTS=/some/share/fractal/folder/artifacts                          755, visible to all users
* JOBDIR=/some/share/fractal/folder/artifacts/workflow_01_job_02          755, visible to all users
* TASKDIR=/some/share/fractal/folder/tasks                                755, visible to all users

Folder owned by user (i.e. test01):
* PROJECT_DIR=/some/share/test01/folder                                   700, but visible to fractal through sudo-cat
* USERJOBDIR=/some/share/test01/folder/tmp_wf_01_job_01   700, but visible to fractal through sudo-cat

Notes:
* PROJECT_DIR must exist (check upon project creation)
* USERJOBDIR must exist (mkdir at the beginning of submit.sh) &#x1F534;

# Workflow execution

### Current version
* fractal writes args.json and submit.sh input.pickle to JOBDIR (and also logs, but that doesn't matter because test01 has no access to workflow.log)

### Future version
* same as current version

### Current version

Steps:
* fractal runs `sudo -u test01 sbatch JOBDIR/submit.sh`
  (OK, since JOBDIR is 777)
* test01 (via sbatch) runs `srun /some/path/python -m cfut JOBDIR/input.pickle`
  (OK, since JOBDIR is 777)
* test01 (via srun) runs `TASKDIR/some/path/python TASKDIR/some/path/task.py --input JOBDIR/args.json`
  (OK, since JOBDIR is 777)
* test01 (via sbatch) writes slurm err/out files to JOBDIR
  (OK, since JOBDIR is 777)

### Future version

`JOBDIR/submit.sh` will look like
```
mkdir -p USERJOBDIR  &#x1F534;
srun /some/path/python -m cfut JOBDIR/input.pickle --output-dir USERJOBDIR
```

Steps:
* fractal runs `sudo -u test01 sbatch JOBDIR/submit.sh`
  (OK, since JOBDIR is 755)
* test01 (via sbatch) runs `mkdir -p USERJOBDIR` &#x1F534;
  (OK, since USERJOBDIR belongs to test01)
* test01 (via sbatch) runs `srun /some/path/python -m cfut JOBDIR/input.pickle --output-dir USERJOBDIR`
  (OK, since JOBDIR is 755 and USERJOBDIR belongs to test01)
* test01 (via srun) runs `TASKDIR/some/path/python TASKDIR/some/path/task.py --input JOBDIR/args.json --output USERJOBDIR/output.json`
  (OK, since JOBDIR is 755 and USERJOBDIR belongs to test01)
* test01 (via sbatch) writes slurm err/out files to USERJOBDIR
  (OK, since USERJOBDIR belongs to test01)

### Current version
* fractal directly (AKA from python) polls existence of JOBDIR/out.pickle (which fractal has access to) every second, and proceeds as soon as the file is there
* if JOBDIR/out.pickle is missing AND squeue polling (taking place e.g. every minute) says job is over, then proceed with error handling

### Future version
* fractal indirectly (AKA subproces.run("sudo -u test01 ls ..")) polls USERJOBDIR/out.pickle every second, and proceeds as soon as the file is there
* as soon as USERJOBDIR/out.pickle exists:
    - trigger copy of all task-related files from USERJOBDIR into JOBDIR (via sudo-cat)
* if  USERJOBDIR/out.pickle is missing AND squeue polling (taking place e.g. every minute) says SLURM job is over, then
    - trigger copy of all task-related files from USERJOBDIR into JOBDIR (via sudo-cat)
    - proceed with error handling

----

# Comment about abstractions

We may say that the proposed update goes one step further towards abstracting out the data exchange between server and user.
**Note**: this is not an abstraction in the sense that we will then be able to just replace one implementation with another one (via a standardized interface, like what we did for concurrent.futures standardized Executor), but rather an abstraction as in "we roughly know which code blocks will need to be replaced with new logic in future versions".

V0 (current):
* server->user is trivial, thanks to 777 folder
* user->server is trivial, thanks to 777 folder

V1 (proposed change)
* server->user is trivial, thanks to 755 folder
* user->server is non-trivial, and requires sudo privileges

V2 (one day)
* Both server->user and user->server are implemented via custom script send_data.py

V3 (one day)
* Both server->user and user->server are implemented with open sockets on both sides, pushing/receiving data back and forth

---

# Open questions:
* Is `local` backend meant as a test for the slurm one, or rather as an independently useful backend? We think that the local backend is important in itself, and not just for testing.
* Based on the previous point, the current proposed solution is to make the backend-independent runner component flexible, and then only implement the proposed change for the slurm backend.
* Number of files in job folders is currently too large, because each task will have something like 7 files (out/err/in.json/out.json/slurm.sh/slurm.out/slurm.err). If a dataset has 400 wells and a workflow has 5 tasks, this means 14k files, which is not acceptable. Ref #474

**Comment by tcompa** [comment]
Minor edit: we can start by not having the `complete` file, and just polling the out.pickle file as it is currently done.

**Comment by tcompa** [comment]
I added &#x1F534; symbols next to steps based on a wrong assumption. Main comment to be updated later, depending on the route we choose.

---

Briefly, a `submit.sh` script like the following (to be used via sbatch) would be flawed:
```bash
# SBATCH --output=/some/non/existing/folder/out

mkdir /some/non/existing/folder
```
The reason is that because `/some/non/existing/folder` does not exist, slurm cannot open the output file right away and will then fail (silently or in an unexpected ways). The mkdir command is never executed.

---

This means that USERJOBDIR needs to be created somewhere else, and this operation needs to take place before the first sbatch execution. The option we are trying now in #475 is to do so with a `sudo -u test01 mkdir /some/non/existing/folder` command (via subprocess).

(all in progress)

**Comment by tcompa** [comment]
I marked another wrong assumption with &#x1F534; (the args json file are written by the user, not by fractal).

**Comment by tcompa** [comment]
For the record, and as a self reminder.

TL;DR we can use (sudo-)cat to read a file, but then write it directly from python with standard methods. This way we reduce (by a tiny bit..) the scope of the operations that are executed via `subprocess.run`.

---

This script
```python
import subprocess
import pickle

data = {"a": 1}
FILE = "old.pickle"
NEWFILE = "new.pickle"

with open(FILE, "wb") as f:
    pickle.dump(data, f)

with open(FILE, "rb") as f:
    print(pickle.load(f))

res = subprocess.run(["cat", FILE], capture_output=True, check=True)
print(res.returncode)
print(res.stdout)
print(res.stderr)

with open(NEWFILE, "wb") as f:
    f.write(res.stdout)

with open(NEWFILE, "rb") as f:
    print(pickle.load(f))
```
produces this output:
```
{'a': 1}
0
b'\x80\x04\x95\n\x00\x00\x00\x00\x00\x00\x00}\x94\x8c\x01a\x94K\x01s.'
b''
{'a': 1}
```

Note that we did not set `encoding="utf-8"` in `subprocess.run`.

(and then adding `sudo -u test01` in front of cat doesn't change anything)

**Comment by tcompa** [comment]
Here is a general summary of the current status of #475.
Briefly: CI is passing, but real-life tests are still needed to verify the expected behavior.

# Updates in `runner` (not backend-specific)

1. All code paths where `workflow_dir` was used are now also using `workflow_dir_user` (to host all files that would be written by the user). The `submit_workflow` interface takes care of defining and creating the two folders; for the local backend they are the same folder, while for the slurm backend they differ and are owned by different users. The user-side one is a subfolder of the project_dir.
2. We added this new folder to the `ApplyWorkflow` model in fractal-common.
3. We removed all references to custom `file_opener`s of folder-creation with 777 mode. The main job dir is now 755.
4. The interface of `call_parallel_task` becomes closer to the `call_single_task` one, in that now they both take a `TaskParameters` argument (instead of a future). This is a minor change, which was useful for point 3 below.
5. `recursive_task_submission` was blocking only for the parallel-task case, while it now becomes blocking for both cases - see
```python
    if this_task.is_parallel:
        # NOTE: call_parallel_task is blocking, i.e. the returned future always
        # has `this_task_future.done() = True`
        this_task_future = call_parallel_task(
            executor=executor,
            task=this_task,
            task_pars_depend=task_pars_depend,
            workflow_dir=workflow_dir,
            workflow_dir_user=workflow_dir_user,
            submit_setup_call=submit_setup_call,
        )
    else:
        # NOTE: executor.submit(call_single_task, ...) is non-blocking, i.e.
        # the returned future may have `this_task_future.done() = False`
        extra_setup = submit_setup_call(
            this_task, task_pars, workflow_dir, workflow_dir_user
        )
        this_task_future = executor.submit(
            call_single_task,
            task=this_task,
            task_pars=task_pars_depend,
            workflow_dir=workflow_dir,
            workflow_dir_user=workflow_dir_user,
            **extra_setup,
        )
        # Wait for the future result (blocking)
        this_task_future.result()                        # NOTE THIS LINE
```

6. Writing the most recent metadata was done as part of `call_single_task` or `call_parallel_task`, but would be weird with the new separate-folder behavior (the two functions are executred by two different users, who have different write permissions). After point 3 above, we could move the two blocks to a single place (the end of `recursive_task_submission`).

# Separate-folders handling in slurm backend

1. From now on, the slurm backend always requires the `slurm_user` to be specified.
3. A `_subprocess_run_as_user.py` module provides function like `_run_command_as_user` or `_mkdir_as_user`, which use `sudo` to impersonate another user and run some commands (typically: `ls`, `mkdir`, `cat`). This is all via `subprocess.run`.
4. The `remote.py` module is now more flexible in handling arbitrary input/output pickle files (together with a little new argparse interface).
5. We subclass `cfut.FileWaitThread` to `FractalFileWaitThread`, to replace the file-existence check with a custom one (we cannot use `os.path.exists`, because the fractal user wouldn't have sufficient permissions.. thus we have to wrap `sudo -u test01 ls ...`)
7. Because of the previous point, we also have to rewrite `cfut.SlurmWaitThread`, so that it inherits from `FractalFileWaitThread`. FIXME: this can be made simpler, with a single subclassing.
8.  We removed all references to custom `file_opener`s of folder-creation with 777 mode.
9. Major update to executor.py to handle the two separate folders. This includes e.g. a new method `FractalSlurmExecutor._copy_files_from_user_to_server`, to be called upon job completion.

**Comment by tcompa** [comment]
Now available in 1.0.5.
